OME005/index.md

```diff
@@ -349,6 +399,119 @@ required.  With the corresponding reader support, this would provide
+  eliminate all uses before adding back.
+
+Once the `CoreMetadata` changes are in place, we can create a
+`SubResolutionFormatReader` to contain a `CoreMetadataList` and to
+fill it from `MetadataRetrieve` when set, or alternatively from
+`setCoreMetadataList` directly.  At this point, it's possible to
+specify all series and resolutions before `setId`.  This work is begun
+on the `format-writer-subresolution` branch.  The writer
+implementation should strictly enforce correct series, resolution and
+plane progression, to ensure correct ordering in the file being
+written.
+
+The TIFF saving code needs updating to allow saving of sub-resolution
+planes in `SUBIFDS`.  I would suggest updating `TiffSaver` to maintain
+a stack of sub-resolutions and to write them in order like libtiff.
+This means it will have to maintain this state:
+
+- `previous` top-level IFD offset (for updating the next pointer
+  tranparently); currently requires specifying up front, which is
+  buggy and the cause of several bugs, causing invalid TIFFs to be
+  written with `next` pointers to nowhere.
+- `SUBIFDS` offset and size and current SubIFD for current IFD.  This
+  will allow automatic update of the parent `SUBIFDS` IFD offset when
+  the current IFD is written, and for all subsequent `SUBIFDS`.  This
+  needs to stack to cater for `SUBIFDS` containing `SUBIFDS`, i.e. a
+  tree structure.
+
+Next, we can make individual readers implement
+`SubResolutionFormatWriter`.  Given the small number of writers, I
+would suggest converting the entire lot for ease of maintenance
+(writers don't need to use subresolutions if they implement the
+interface).  Then we can update `OMETiffWriter` to write pyramids.
+Given the current series and current resolution information in the
+inherited writer implementation, the only alteration is to `saveBytes`
+to set the correct `SUBIFDS` size when writing out the full resolution
+plane, then writing each resolution sequentially and updating
+`SUBIFDS` accordingly (if `TiffSaver` is made stateful this can be
+automatic).  The writer should set `NEWSUBFILETYPE` appropriately
+to full resolution or reduced-resolution image.  Also bitwise OR PAGE if
+the image is a multi-plane series.
+
+Lastly, update `ImageConverter` to set the resolutions in
+`CoreMetadataList` and use `setResolution` in addition to `setSeries`
+to loop over each resolution as well as each series and transfer all
+the resolution levels.  The `imageconverter-noflat` and
+`ometiff-pyramid-writer` branches implement most of the needed logic.
+
+## Outstanding questions
+
+- Do we need to store metadata to describe alignment offsets between
+  sub-resolution levels when the size difference between levels
+  results in pixels not aligned on pixel boundaries, for example with
+  non-power-of-two reductions.  How is this handled by the existing
+  pyramid file formats? (Question from 2018 OME annual meeting in
+  Dundee.)
```

Comment:
We should also address whether or not there is any enforcement of the XY sizes of each resolution.  In particular, what happens if someone tries to write a pyramid of e.g. 1024x1024 => 512x512 => 200x200?